packages/framework/tests/Unit/Component/Context/ContextResolverTest.php

```diff
@@ -0,0 +1,177 @@
+    {
+        $context = new TestContextA(shouldMatch: true);
+        $resolver = new ContextResolver([$context]);
+
+        // Test that context was registered correctly by checking if it matches
+        $this->assertTrue($resolver->isCurrentContext(TestContextA::class));
+    }
+
+    public function testConstructorThrowsExceptionForDuplicateContexts(): void
+    {
+        $context1 = new TestContextA();
+        $context2 = new TestContextA();
+
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage('Context with identifier "' . TestContextA::class . '" is already registered');
+
+        new ContextResolver([$context1, $context2]);
+    }
+
+    public function testContextMatching(): void
+    {
+        $context1 = new TestContextA(shouldMatch: false);
+        $context2 = new TestContextB(shouldMatch: true);
+        $context3 = new TestContextC(shouldMatch: true);
+
+        $resolver = new ContextResolver([$context1, $context2, $context3]);
+
+        // Check which contexts match
+        $this->assertFalse($resolver->isCurrentContext(TestContextA::class));
+        $this->assertTrue($resolver->isCurrentContext(TestContextB::class));
+        $this->assertTrue($resolver->isCurrentContext(TestContextC::class));
+    }
+
+    public function testContextWithRequiredContexts(): void
+    {
+        // Create extended context that requires TestContextB
+        $extendedContext = new TestContextA(shouldMatch: true, requiredContexts: [TestContextB::class]);
+        $baseContext = new TestContextB(shouldMatch: true);
+        $defaultContext = new TestContextC(shouldMatch: true);
+
+        $resolver = new ContextResolver([$extendedContext, $baseContext, $defaultContext]);
+
+        // All contexts should match since their dependencies are satisfied
+        $this->assertTrue($resolver->isCurrentContext(TestContextA::class));
+        $this->assertTrue($resolver->isCurrentContext(TestContextB::class));
+        $this->assertTrue($resolver->isCurrentContext(TestContextC::class));
+    }
+
+    public function testContextWithNonMatchingRequiredContext(): void
+    {
+        // Required context that doesn't match
+        $baseContext = new TestContextB(shouldMatch: false);
+        // Context that requires the non-matching context
+        $extendedContext = new TestContextA(shouldMatch: true, requiredContexts: [TestContextB::class]);
+
+        $resolver = new ContextResolver([$baseContext, $extendedContext]);
+
+        // Extended context shouldn't match because required base context doesn't match
+        $this->assertFalse($resolver->isCurrentContext(TestContextA::class));
+        $this->assertFalse($resolver->isCurrentContext(TestContextB::class));
+    }
+
+    public function testContextMatchingResultsAreCached(): void
+    {
+        $matchesCallCount = 0;
+
+        $context = new class($matchesCallCount) extends AbstractContext {
```

Comment:
This could also be achieved more easily with mocking and expecting the method call count, IMHO

Reply:
Mocking AbstractContext is not possible because of using class strings and also validating them. Mock is just an anonymous class that doesn't physically exist. So, validation of context classes will not work.

Also, we would need to "hack" PHPStan. Because of this, creating concrete classes is probably the best way to test this piece of code.